// @sudoku/stores/history/state.js
import { get, writable } from 'svelte/store';

// 栈与计数状态集中管理，便于被 API 组合调用
const branchBackTimeSteps = [];
const branchBackTimes = writable(0);
const undoStack = [];
const redoStack = [];
const undoCount = writable(0);
const redoCount = writable(0);

export const historyState = {
  branchBackTimeSteps,
  branchBackTimes,
  undoStack,
  redoStack,
  undoCount,
  redoCount,

  resetBranches: () => {
    branchBackTimeSteps.length = 0;
    branchBackTimes.set(0);
  },

  addBranch: (timeStep) => {
    branchBackTimeSteps[get(branchBackTimes)] = timeStep;
    branchBackTimes.update((v) => v + 1);
  },

  popBranch: () => {
    const count = get(branchBackTimes);
    if (count <= 0) return undefined;
    branchBackTimes.update((v) => Math.max(0, v - 1));
    return branchBackTimeSteps[count - 1];
  },

  peekBranch: () => {
    const count = get(branchBackTimes);
    if (count <= 0) return undefined;
    return branchBackTimeSteps[count - 1];
  },

  pushUndo: (action) => {
    undoStack.push(action);
    undoCount.set(undoStack.length);
  },

  popUndo: () => {
    const action = undoStack.pop();
    undoCount.set(undoStack.length);
    return action;
  },

  clearRedo: () => {
    redoStack.length = 0;
    redoCount.set(0);
  },

  pushRedo: (action) => {
    redoStack.push(action);
    redoCount.set(redoStack.length);
  },

  popRedo: () => {
    const action = redoStack.pop();
    redoCount.set(redoStack.length);
    return action;
  },
};
